File: AndroidArchitecture/domain/src/main/java/tisoul/dev/domain/domain/usecase/impl/GetUserProfileUseCase.java
package tisoul.dev.domain.domain.usecase.impl;


import javax.inject.Inject;

import io.reactivex.Observable;
import tisoul.dev.domain.domain.usecase.GetUserProfile;
import tisoul.dev.domain.model.UserProfile;
import tisoul.dev.domain.repository.ProfileRepository;
import tisoul.dev.domain.repository.model.ResultCallback;

public class GetUserProfileUseCase
        implements GetUserProfile {

    private ProfileRepository profileRepository;

    @Inject
    public GetUserProfileUseCase(ProfileRepository profileRepository) {
        this.profileRepository = profileRepository;
    }

    @Override
    public Observable<UserProfile> execute(Void aVoid) {
        return Observable
                .create(emitter -> profileRepository
                        .getProfile(
                                new ResultCallback<UserProfile>() {
                                    @Override
                                    public void onSuccess(UserProfile userProfile) {
                                        emitter.onNext(userProfile);
                                        emitter.onComplete();
                                    }

                                    @Override
                                    public void onError(Throwable throwable) {
                                        emitter.onError(throwable);
                                    }
                                }));
    }
}

File: AndroidArchitecture/domain/src/main/java/tisoul/dev/domain/domain/usecase/impl/GetUserProfileUseCase.kt
package tisoul.dev.domain.domain.usecase.impl


import tisoul.dev.domain.domain.usecase.GetUserProfile
import tisoul.dev.domain.repository.ProfileRepository
import javax.inject.Inject

class GetUserProfileUseCase
@Inject constructor(
        private val profileRepository: ProfileRepository)
    : GetUserProfile {

    override fun execute(p: Boolean)
            = profileRepository.getProfile()
}
